### DISCUSSION TITLE
Setting up `mini.ai` alongside `nvim-treesitter-textobjects` and setting up treesitter fallbacks

### DISCUSSION THREAD

**Discussion by simonmandlik** [nvim-mini/mini.nvim#970] [Q&A] [answered]
Hi!

I'm trying to setup treesitter-based textobjects using `mini.ai` and `nvim-treesitter-textobjects`. I really like the `g]` `g[` jumps, "counts", and next/last extension from `mini.ai`, on the other hand, I can't imagine living without `include_surrounding_whitespace` and also `selection_modes` options from `nvim-treesitter-textobjects`. I was wondering if anyone came up with a setup that gets the best of both worlds.

I was also wondering if it is possible to setup "fallbacks" when treesitter isn't available. In buffers with treesitter support, I would for example like to use query from treesitter for calls, whereas in non-treesitter buffers, I would like to fallback to regex-based definitions.

Thank you

**Comment by echasnovski** [comment, marked as answer]
Hello!

> ... I can't imagine living without `include_surrounding_whitespace` and also `selection_modes` options from `nvim-treesitter-textobjects`. I was wondering if anyone came up with a setup that gets the best of both worlds.

Both of those options don't look like adding too much of a feature set, to be honest.

`include_surrounding_whitespace` *might* be useful, but mostly feels like an ad hoc patch that fixes the tree-sitter grammar implementation rather that the problem itself.

The `selection_modes` seems available in 'mini.ai' via [`vis_mode` field of a textobject specification](https://github.com/echasnovski/mini.nvim/blob/ebc8cb0c1f73ee6c53e0560957a37a2307a70f98/doc/mini-ai.txt#L192). But it seems to only be present for manual specifications, while `gen_spec.treesitter()` can not forward this information. I am not against allowing all `gen_spec` entries to forward this information per se, but it looks like needing extra testing/documentation for what I perceive as minor gain. Mostly because it would benefit Operator-pending mode mappings; Visual mode can be initiated manually with `v`, `V`, or `<C-v>` and it will stay that after textobject selection.

> I was also wondering if it is possible to setup "fallbacks" when treesitter isn't available. In buffers with treesitter support, I would for example like to use query from treesitter for calls, whereas in non-treesitter buffers, I would like to fallback to regex-based definitions.

Not quite fallbacks, but you can disable 'mini.ai' textobject per buffer. For this, create a `vim.b.miniai.config` buffer-local variable and set target textobject identifier to `false` ([like this](https://github.com/echasnovski/mini.nvim/blob/ebc8cb0c1f73ee6c53e0560957a37a2307a70f98/doc/mini-ai.txt#L455-L456)).

Besides that you'd have to write your own textobject in function form.

**Comment by simonmandlik** [comment]
Thanks!
